File: backend/utils/text_cleaner.py
"""
Text cleaning utilities for resume content.
Normalises whitespace and removes extraction noise
without destroying meaningful formatting.
"""

import re


def clean_resume_text(text: str) -> str:
    """
    Clean extracted resume text.

    Steps:
    1. Normalise Windows line endings to Unix.
    2. Remove NULL bytes and control characters.
    3. Collapse runs of whitespace that are not newlines.
    4. Remove lines that are purely noise (long repeated punctuation).
    5. Collapse more than two consecutive blank lines into two.
    6. Strip leading/trailing whitespace from the whole document.
    """
    if not text:
        return ""

    # 1. Normalise line endings
    text = text.replace("\r\n", "\n").replace("\r", "\n")

    # 2. Remove NULL bytes and other control characters (keep tab, newline)
    text = re.sub(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]", " ", text)

    # 3. Collapse horizontal whitespace (spaces/tabs) within a line
    lines = text.split("\n")
    cleaned_lines = []
    for line in lines:
        line = re.sub(r"[ \t]+", " ", line).strip()
        cleaned_lines.append(line)

    # 4. Remove lines that are purely repeated punctuation / decoration
    filtered_lines = []
    for line in cleaned_lines:
        # Skip lines that are only punctuation/symbols repeated 3+ times
        if re.fullmatch(r"[^\w\s]{3,}", line):
            continue
        # Skip lines that are only underscores, dashes, equals (horizontal rules)
        if re.fullmatch(r"[-_=*#~]{3,}", line):
            continue
        filtered_lines.append(line)

    # 5. Collapse more than two consecutive blank lines
    result_lines: list[str] = []
    blank_count = 0
    for line in filtered_lines:
        if line == "":
            blank_count += 1
            if blank_count <= 2:
                result_lines.append(line)
        else:
            blank_count = 0
            result_lines.append(line)

    return "\n".join(result_lines).strip()


def is_text_meaningful(text: str, min_words: int = 50) -> bool:
    """
    Return True if the text contains enough content to be analysed.
    A resume with fewer than `min_words` words is considered too short.
    """
    words = text.split()
    return len(words) >= min_words

File: backend/resumeiq-api/src/main/java/com/resumeiq/util/TextCleaner.java
package com.resumeiq.util;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class TextCleaner {

    private static final Pattern NULL_BYTES_PATTERN = Pattern.compile("[\\x00-\\x08\\x0b\\x0c\\x0e-\\x1f\\x7f]");
    private static final Pattern WHITESPACE_PATTERN = Pattern.compile("[ \\t]+");
    private static final Pattern PUNCTUATION_REPEATED_PATTERN = Pattern.compile("[^\\w\\s]{3,}");
    private static final Pattern HORIZONTAL_RULE_PATTERN = Pattern.compile("[-_=*#~]{3,}");

    public static String cleanResumeText(String text) {
        if (text == null || text.isEmpty()) {
            return "";
        }

        // 1. Normalise line endings
        text = text.replace("\r\n", "\n").replace("\r", "\n");

        // 2. Remove NULL bytes and other control characters (keep tab, newline)
        text = NULL_BYTES_PATTERN.matcher(text).replaceAll(" ");

        // 3. Collapse horizontal whitespace (spaces/tabs) within a line
        String[] lines = text.split("\n");
        List<String> cleanedLines = new ArrayList<>();
        for (String line : lines) {
            line = WHITESPACE_PATTERN.matcher(line).replaceAll(" ").trim();
            cleanedLines.add(line);
        }

        // 4. Remove lines that are purely repeated punctuation / decoration
        List<String> filteredLines = new ArrayList<>();
        for (String line : cleanedLines) {
            if (PUNCTUATION_REPEATED_PATTERN.matcher(line).matches()) {
                continue;
            }
            if (HORIZONTAL_RULE_PATTERN.matcher(line).matches()) {
                continue;
            }
            filteredLines.add(line);
        }

        // 5. Collapse more than two consecutive blank lines
        List<String> resultLines = new ArrayList<>();
        int blankCount = 0;
        for (String line : filteredLines) {
            if (line.isEmpty()) {
                blankCount++;
                if (blankCount <= 2) {
                    resultLines.add(line);
                }
            } else {
                blankCount = 0;
                resultLines.add(line);
            }
        }

        return String.join("\n", resultLines).trim();
    }

    public static boolean isTextMeaningful(String text, int minWords) {
        if (text == null || text.trim().isEmpty()) {
            return false;
        }
        String[] words = text.split("\\s+");
        return words.length >= minWords;
    }
}
